tools/deployment_tool/contract_registry.js

```diff
@@ -0,0 +1,127 @@
+const {
+    IncrementingAddressGenerator,
+    IncrementingNonceAddressGenerator,
+} = require('./address_generators');
+
+/**
+ * A contract registry that allows for registering contracts and planning deployment.
+ */
+class ContractRegistry {
+    constructor() {
+        this.contracts = [];
+
+        this.addContract = this.addContract.bind(this);
+        this.toParityGenesisAccounts = this.toParityGenesisAccounts.bind(this);
+        this.toLiveTransactionObjects = this.toLiveTransactionObjects.bind(this);
+    }
+
+    /**
+     * Add a contract to the registry.
+     *
+     * @param {Contract} contract Contract to add to the registry.
+     */
+    addContract(contract) {
+        this.contracts.push(contract);
+    }
+
+    /**
+     * Add multiple contracts to the registry. See {@link ContractRegistry#addContract}.
+     *
+     * @param {Array.<Contract>} contracts Contracts to add to the registry.
+     */
+    addContracts(contracts) {
+        contracts.forEach(contract => this.addContract(contract));
+    }
+
+    /**
+     * Generate the "accounts" object for a parity chainspec for all the contracts have been added.
+     * For that the { "0x...": { "constructor": "0x..." } } version of account intialization
+     * which is currently exclusive to parity is used.
+     *
+     * See {@link https://wiki.parity.io/Chain-specification} for more details on the
+     * parity chainspec format.
+     *
+     * @param {Object} options.addressGenerator Address generator to use.
+     *                 Defaults to a IncrementingAddressGenerator.
+     *
+     * @returns {object} The "accounts" section for a parity chainspec.
+     */
+    toParityGenesisAccounts(options = {}) {
+        const defaultOptions = {
+            addressGenerator: new IncrementingAddressGenerator(),
+        };
+        const mergedOptions = Object.assign(defaultOptions, options);
+
+        // prepare contracts by ordering and generating addresses
+        const { addressGenerator } = mergedOptions;
+        this._orderContracts();
+        this.contracts.forEach(contract => contract.setAddress(addressGenerator));
+        this.contracts.forEach(contract => contract.instantiate());
+
+        const genesisAccounts = {
+        };
+
+        this.contracts
+            .forEach((contract) => {
+                const address = contract.getAddress();
+                const constructor = contract.constructorData;
+
+                genesisAccounts[address] = {
+                    balance: '0',
+                    constructor,
+                };
+            });
+
+        return genesisAccounts;
+    }
+
+    /**
+     * Generate transaction objects that can be passed as an argument
+     * to `web3.eth.sendTransaction()`.
+     * This allows for deployment on a live network.
```

Comment:
It would help to reference `toParityGenesisAccounts` and `toLiveTransactionObjects` to each other I think. What do you think?

Reply:
Done!